brave-cxf3/src/main/java/com/github/kristofa/brave/cxf3/BraveCxfExecutorService.java

```diff
@@ -0,0 +1,157 @@
+import com.github.kristofa.brave.ServerSpanThreadBinder;
+import org.apache.cxf.message.Message;
+import org.apache.cxf.phase.PhaseInterceptorChain;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.*;
+
+/**
+ * Wraps around {@link ExecutorService} and passes {@link ServerSpan}
+ * from cxf {@link org.apache.cxf.message.Exchange} context (if exists)
+ * to current {@link ServerSpanThreadBinder} during execution.
+ */
+public class BraveCxfExecutorService implements ExecutorService, Closeable {
+
+  private final ExecutorService delegate;
+  private final ServerSpanThreadBinder serverSpanThreadBinder;
+
+  public static BraveCxfExecutorService wrap(ExecutorService delegate, Brave brave) {
+    return new BraveCxfExecutorService(delegate, brave);
+  }
+
+  BraveCxfExecutorService(ExecutorService delegate, Brave brave) {
+    this.delegate = delegate;
+    this.serverSpanThreadBinder = brave.serverSpanThreadBinder();
+  }
+
+  @Override
+  public void close() throws IOException {
+    shutdown();
+  }
+
+  @Override
+  public void execute(Runnable command) {
+    delegate.execute(wrap(command));
+  }
+
+  @Override
+  public void shutdown() {
+    delegate.shutdown();
+  }
+
+  @Override
+  public List<Runnable> shutdownNow() {
+    return delegate.shutdownNow();
+  }
+
+  @Override
+  public boolean isShutdown() {
+    return delegate.isShutdown();
+  }
+
+  @Override
+  public boolean isTerminated() {
+    return delegate.isTerminated();
+  }
+
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    return delegate.awaitTermination(timeout, unit);
+  }
+
+  @Override
+  public <T> Future<T> submit(Callable<T> task) {
+    return delegate.submit(wrap(task));
+  }
+
+  @Override
+  public <T> Future<T> submit(Runnable task, T result) {
+    return delegate.submit(wrap(task), result);
+  }
+
+  @Override
+  public Future<?> submit(Runnable task) {
+    return delegate.submit(wrap(task));
+  }
+
+  @Override
+  public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks) throws InterruptedException {
+    return delegate.invokeAll(wrap(tasks));
+  }
+
+  @Override
+  public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit) throws InterruptedException {
+    return delegate.invokeAll(wrap(tasks), timeout, unit);
+  }
+
+  @Override
+  public <T> T invokeAny(Collection<? extends Callable<T>> tasks) throws InterruptedException, ExecutionException {
+    return delegate.invokeAny(wrap(tasks));
+  }
+
+  @Override
+  public <T> T invokeAny(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+    return delegate.invokeAny(wrap(tasks), timeout, unit);
+  }
+
+  private <T> T getFromExchange(String key) {
```

Comment:
I don't think so, because in async scenario RequestFilter would be executed on Thread1 and Response will be finished on Thread2, thus T1 wouldn't be cleaned.
Eventually we could get rid of BraveAsyncJaxRsThreadCleaner and BraveCxfExecutorService could pass value from T1 to T2, and finally clear T1.